Adopt GB18030-2022 into GBK more fully

annevk opened the pull request:
#### c8a687829285dd572528fca4fd8aa9abf4acff09
<pre>
Adopt GB18030-2022 into GBK more fully
<a href="https://bugs.webkit.org/show_bug.cgi?id=279903">https://bugs.webkit.org/show_bug.cgi?id=279903</a>

Reviewed by Alex Christensen.

For GBK and gb18030 we have used the same backing table for quite a
while now. This backing table was updated to account for GB18030-2022
at some point and this impacted GBK as well.

However, the encoder side table was kept disabled for GBK, despite it
actually allowing GBK to be more compatible with its former self.

<a href="https://github.com/whatwg/encoding/pull/336">https://github.com/whatwg/encoding/pull/336</a> now standardizes the
behavior that GBK and gb18030 are to remain aligned in these matters
and this change implements that.

The corresponding tests are from this PR:
<a href="https://github.com/web-platform-tests/wpt/pull/48240">https://github.com/web-platform-tests/wpt/pull/48240</a>

* LayoutTests/imported/w3c/web-platform-tests/encoding/legacy-mb-schinese/gbk/gbk-decoder.any.js:
* LayoutTests/imported/w3c/web-platform-tests/encoding/legacy-mb-schinese/gbk/gbk-encoder-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/encoding/legacy-mb-schinese/gbk/gbk-encoder.html:
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::gb18030AsymmetricEncode):
(PAL::gbEncodeShared):

Canonical link: <a href="https://commits.webkit.org/283987@main">https://commits.webkit.org/283987@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fa7ebc3d618d2515087052d72a36f6cbd7c7801e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67610 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46989 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20242 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71664 "Built successfully") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69728 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54787 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18551 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54143 "Passed tests") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70677 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43119 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58466 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34615 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39783 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15871 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17109 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61750 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16212 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73359 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11569 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15532 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61589 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11604 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58535 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61643 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15112 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9448 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3062 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42795 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43873 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45057 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43612 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
```diff
@@ -899,7 +899,7 @@ static const GB18030EncodeIndex& gb18030EncodeIndex()
 // The 2-byte values are handled correctly by values from gb18030()
 // but these need to be exceptions from gb18030Ranges().
-static std::optional<uint32_t> gb18030AsymmetricEncode(char32_t codePoint)
+static std::optional<uint16_t> gb18030AsymmetricEncode(UChar codePoint)
```
Please carefully review these changes. As far as I can tell it's never called with anything but a UChar, so not sure why we'd take a char32_t. And it only needs to return two bytes.

achristensen07 replied:
Yes, this is fine.

achristensen07 commented on LayoutTests/imported/w3c/web-platform-tests/encoding/legacy-mb-schinese/gbk/gbk-decoder.any.js:
```diff
@@ -1,3 +1,5 @@
+// Additional tests can be found in ../gb18030/gb18030-decoder.any.js
```
This is true, but no tests for gbk are in the gb18030 directory, right?  I don't think we should add this comment.

annevk replied:
The gbk decoder is tested there. I was surprised by that as well, but it makes sense as the specification defines them to be the same:
```
const decode = (input, output, desc) => {
  test(function () {
    for (const encoding of ["gb18030", "gbk"]) {
      assert_equals(
        new TextDecoder(encoding).decode(new Uint8Array(input)),
        output,
      );
    }
  }, "gb18030 decoder: " + desc);
};
```
That's also why I don't feel the need to add additional decoder tests here for `gbk`.

annevk committed:
Adopt GB18030-2022 into GBK more fully
https://bugs.webkit.org/show_bug.cgi?id=279903

Reviewed by Alex Christensen.

For GBK and gb18030 we have used the same backing table for quite a
while now. This backing table was updated to account for GB18030-2022
at some point and this impacted GBK as well.

However, the encoder side table was kept disabled for GBK, despite it
actually allowing GBK to be more compatible with its former self.

https://github.com/whatwg/encoding/pull/336 now standardizes the
behavior that GBK and gb18030 are to remain aligned in these matters
and this change implements that.

The corresponding tests are from this PR:
https://github.com/web-platform-tests/wpt/pull/48240

* LayoutTests/imported/w3c/web-platform-tests/encoding/legacy-mb-schinese/gbk/gbk-decoder.any.js:
* LayoutTests/imported/w3c/web-platform-tests/encoding/legacy-mb-schinese/gbk/gbk-encoder-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/encoding/legacy-mb-schinese/gbk/gbk-encoder.html:
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::gb18030AsymmetricEncode):
(PAL::gbEncodeShared):

Canonical link: https://commits.webkit.org/283987@main
